### PR TITLE
cache: Column option -k/--key

### PR DESCRIPTION
**Pull request by lpla**
Now you can provide column(s) key to use as the deduplication string using -k/--key in the `cache` program. If several, separate them using commas: `cat foo.txt | cache --key 3,4 test.sh`

### CODE REVIEW

File: preprocess/cache_main.cc [hunk 1]
```diff
@@ -50,14 +60,65 @@ struct QueueEntry {
...
+      token_args = s_args.substr(0, pos_args);
+      std::string::size_type sz;
+      int int_token_args = std::stoi (token_args, &sz);
+      columns_args.push_back(int_token_args);
+      s_args.erase(0, pos_args + 1);
+    }
+    std::string::size_type sz;
+    int last_int_token_args = std::stoi (s_args, &sz);
+    columns_args.push_back(last_int_token_args);
+
```
Comment (kpu) [inline preprocess/cache_main.cc]: See `ParseFields` in `fields.hh`.  There's already a parser for this and it supports `-`.

File: preprocess/cache_main.cc [hunk 2]
```diff
@@ -50,14 +60,65 @@ struct QueueEntry {
...
+        }
+      }
+      // If any column number is out of boundaries (or --key is not provided, so uses default -1), use the whole line as data for mmh
+      if (columns.size() < max || min <= 0){
+        entry.first = util::MurmurHashNative(l.data(), l.size());
+      }
+      else{
+        std::string hash_string;
+        for (int i : columns_args){
+          hash_string += columns.at(i-1);
```
Comment (kpu) [inline preprocess/cache_main.cc]: Don't concatenate strings then hash. Chain hashes using the seed argument.

File: preprocess/cache_main.cc [hunk 3]
```diff
@@ -50,14 +60,65 @@ struct QueueEntry {
...
+      s_args.erase(0, pos_args + 1);
+    }
+    std::string::size_type sz;
+    int last_int_token_args = std::stoi (s_args, &sz);
+    columns_args.push_back(last_int_token_args);
+
     for (StringPiece l : util::FilePiece(STDIN_FILENO)) {
-      entry.first = util::MurmurHashNative(l.data(), l.size());
+      // Split content columns, if given using --key option, into a string vector
+      std::vector<std::string> columns;
```
Comment (kpu) [inline preprocess/cache_main.cc]: Don't use `std::string` when you can use `StringPiece`

File: preprocess/cache_main.cc [hunk 4]
```diff
@@ -50,14 +60,65 @@ struct QueueEntry {
...
+ 
+      size_t pos = 0;
+      std::string token;
+      std::string s (l.data(), l.size());
+      while ((pos = s.find(options.field_separator)) != std::string::npos) {
+        token = s.substr(0, pos);
+        columns.push_back(token);
+        s.erase(0, pos + options.field_separator.length());
```
Comment (kpu) [inline preprocess/cache_main.cc]: O(n^2).  Could be O(n) with StringPiece.

File: preprocess/cache_main.cc [hunk 5]
```diff
@@ -50,14 +60,65 @@ struct QueueEntry {
...
+      // If any column number is out of boundaries (or --key is not provided, so uses default -1), use the whole line as data for mmh
+      if (columns.size() < max || min <= 0){
+        entry.first = util::MurmurHashNative(l.data(), l.size());
+      }
+      else{
+        std::string hash_string;
+        for (int i : columns_args){
+          hash_string += columns.at(i-1);
+        }
+        entry.first = util::MurmurHashNative(hash_string.c_str(), hash_string.size());
```
Comment (kpu) [inline preprocess/cache_main.cc]: I think this chunk can be replaced by `RangeFields` or `IndividualFields` in `fields.hh`

File: preprocess/cache_main.cc [hunk 6]
```diff
@@ -102,22 +163,47 @@ void Output(util::UnboundedSingleQueue<QueueEntry> &queue, util::scoped_fd &proc
...
+  if (s1 == "-k" || s1 == "-t" || s1 == "--key" || s1 == "--field_separator"){
+    skip_args += 2;
+  }
+  if (argc > 3){
+    std::string s3(argv[3]);
+    if (s3 == "-k" || s3 == "-t" || s3 == "--key" || s3 == "--field_separator"){
+      skip_args += 2;
+    }
   }
+  char *partArgs[skip_args];
```
Comment (kpu) [inline preprocess/cache_main.cc]: This is gcc-only code.  Could use a vector.

File: preprocess/cache_main.cc [hunk 7]
```diff
@@ -102,22 +163,47 @@ void Output(util::UnboundedSingleQueue<QueueEntry> &queue, util::scoped_fd &proc
...
   const std::size_t kFlushRate = 4096;
-  if (argc < 2) {
-    std::cerr << "Acts as a cache around another program processing one line in, one line out from stdin to stdout.\n"
-      "Usage: " << argv[0] << " slow arguments_to_slow\n";
-    return 1;
+  
+  // Take into account the number of arguments given to `cache` to delete them from the argv provided to Launch function
+  Options opt;
+  int skip_args = 1;
+  std::string s1(argv[1]);
```
Comment (kpu) [inline preprocess/cache_main.cc]: We're converting a `char *` to a `std::string` just so we can compare it?  `strcmp`.  This is messy but I get why you are doing it.

**Comment by lpla** [comment]
Thank you for your code review. Please, check latest commit to see if everything is correct now.

**Pull request merged**